### ISSUE TITLE
Document inlineCallbacks

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7343]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)| @puzzlement reported|
|-|-|
|Trac ID|trac#1009|
|Type|defect|
|Created|2005-04-30 05:55:42Z|
|Branch|https://github.com/twisted/twisted/tree/inlineCallbacks-howto-1009-2|

Attachments:

* [comments.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ab6/ab68fc51497db0ca21119091a92e418e8a27984f/52c1531ad77faf769144850443ace4c3c191418a.txt) (1287 bytes) - added by hypatia on 2006-04-09 01:58:04Z - Alex Levy's examples with his comments.
* [defgen.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ab6/ab68fc51497db0ca21119091a92e418e8a27984f/47feadd630e43c0f06b51aa3603337d0585237ba.txt) (9715 bytes) - added by edsuom on 2006-06-14 05:26:19Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1009 1009
type__defect defect
reporter__hypatia hypatia
priority__high high
milestone__ 
branch__branches_inlineCallbacks_howto_1009_2 branches/inlineCallbacks-howto-1009-2
branch_author__tomprince__glyph tomprince, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1114840542000000 1114840542000000
changetime__1410216167440483 1410216167440483
version__ 
owner__glyph glyph
cc__radix cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
http://mesozoic.geecs.org/cogito/archives/000160.html suggests that defgen
really needs a prominent place in the Deferred documentation.

There's one open question: how prominent a place in the documentation? When
should people be encouraged to use defgen, and when should they be encouraged to
use a more traditional callback setup?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
As the author of defgen, I have a superior understanding of the subtleties 
involved in deciding when to use it.

That is to say, I don't know.

In my current work project, we haven't used it at all. I'm not really sure why. 
We didn't have many places in the project where we dealt with many Deferreds in 
a single method that we wanted serialized. I'm not sure if that's the end of it, 
but that's usually when I think to use defgen. However, thinking back, we 
probably could  have slightly improved the look of the code where we have 2 or 3 
callbacks attached to a Deferred. It would have at least moved those all into a 
single method.

I'm not sure if the stuff I'm saying here is useful at all, maybe it's even 
misleading. I'll try to have a sit down and think about this sometime soon.

As an aside, for my next work project, I'm looking forward to stabilizing trying 
out gthreadless in production....
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Just a note.

scratch that about gthreadless. However, also note that we did eventually use
defgen quite a bit in our work projects. It turned out pretty well, but error
reporting still sucks a lot in certain cases.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="edsuom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|edsuom set status to `assigned`|
|-|-|

I like defgen and have a use case for it, in a BNF terminal-evaluator project I'm working on:

```python
def evaluator(*inputStates):
    def function():
        for inputState in inputStates:
            wfd = waitForDeferred(self.run(sentence, state))
            yield wfd
            resultState = wfd.getResult()
            results.append(f(inputState, resultState))
    function = deferredGenerator(function)

    return function()
```

I'll include this plus Levy's more advanced example in the docs.

**Comment by twisted-trac** [comment]
|<img alt="edsuom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|edsuom set status to `closed`|
|-|-|

I wrote a tutorial document that is attached and also available as the "Deferred Generators" section of my "WinDictator Internals" [document](http://foss.eepatents.com/WinDictator/wiki/Internals).

I suggest just pasting the source right into a wiki page of the Twisted site, or using ```wget``` to grab the HTML and put it in the regular docs, or both.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @thijstriemstra<br>@exarkun set status to `new`|
|-|-|

This should be a howto and it should cover `inlineCallbacks` in addition to `deferredGenerator` and it should use some simple Deferred API like `getPage`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

[#1821](https://github.com/twisted/twisted/issues/1821) was a duplicate of this.  There's a bunch of stuff attached to that ticket which should almost certainly be incorporated into a resolution to this ticket.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

[#5177](https://github.com/twisted/twisted/issues/5177) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="santagada's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|santagada commented|
|-|-|

From the duplicate bug because I think it is very important:

"There should be a way to show inlineCallbacks to new users to show that you can do asynchronous communication without using callbacks everywhere. In the last pycon(2011) guido himself told that he doesn't understand callback based code and that he thinks the future of async in python is using generators, so I really think this is something the twisted community should talk about."

One thing we may do is ask one of the bloggers that posted about this to lend the code or to rewrite it as a howto, but still in the rest of the docs there should be pointers to inlineCallbacks, not just a separate doc about it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Don't bother documenting `deferredGenerator`, instead we'll deprecate it ([#6044](https://github.com/twisted/twisted/issues/6044)).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince commented|
|-|-|

(In [37454]) Branching to inlineCallbacks-howto-1009.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince commented|
|-|-|

(In [37455]) Apply deferredgenerator-howto-1821.patch from thijs.

Refs [#7343](https://github.com/twisted/twisted/issues/7343), [#1821](https://github.com/twisted/twisted/issues/1821).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince commented|
|-|-|

I've applied the patch from [#1821](https://github.com/twisted/twisted/issues/1821) (which was a lore translation of [[defgen.txt](None/defgen.txt) defgen.txt]) and applied a minimal first pass at switching to documenting `inlineCallbacks` instead.

Review comments from [ticket:1821#comment:6 [#1821](https://github.com/twisted/twisted/issues/1821)]:
 -    The use of the word intuitive in the introduction definitely has to go.
 -    Using reactor.callLater is often confusing to people new to Deferreds. There should be some other event source here - eg, twisted.web.client.getPage
 -    I'm not sure about the use of decorator syntax here. It should probably be replaced with manual decoration, although it may make sense to include at least one example using decorator syntax to help people who can't quite make the connection on their own. The Twisted coding standard forbids decorator syntax.
 -    Only one of the code samples in this document can actually be run usefully. Runnable code is worth a lot more than dead examples.
 -    The document needs to introduce the concepts which it will cover, explain what readers need to already be familiar with, and should include links to other related documents at the end.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

(In [43007]) Branching to inlineCallbacks-howto-1009-2.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @dstufft|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

Completely rewritten prose documentation, in the introductory docs, featuring a lot less verbiage.  (Also in ReST now because we already had lore's funeral)

**Comment by twisted-trac** [comment]
|[<img alt="dstufft's avatar" src="https://avatars.githubusercontent.com/u/145979?s=50" width="50" height="50">](https://github.com/dstufft)<a name="note_22"></a>|@dstufft commented|
|-|-|

Question, are there gotchas or problems from using inlineCallbacks that someone should be aware of?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

Replying to [dstufft](#note_22):
> Question, are there gotchas or problems from using inlineCallbacks that someone should be aware of?

You mean aside from "Concurrency is incredibly confusing and human brains can't really reason about it effectively?"

No, not really.

I wanted to include something about how `yield` suspends execution and so you have to worry about mutual exclusion and critical sections, but that's a very complicated topic to address in introductory documentation.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph commented|
|-|-|

We could potentially have a separate ticket for the Deferred reference docs, but this ticket is like "`inlineCallbacks` should be documented _at all_".

**Comment by twisted-trac** [comment]
|[<img alt="dstufft's avatar" src="https://avatars.githubusercontent.com/u/145979?s=50" width="50" height="50">](https://github.com/dstufft)<a name="note_25"></a>|@dstufft commented|
|-|-|

Yea, that sounds more like something that should be in an advanced topic documentation linked from the introduction as a "more info" kind of thing. I've only done a quick once over, I'll go over it in more depth later this evening.

**Comment by twisted-trac** [comment]
|[<img alt="dstufft's avatar" src="https://avatars.githubusercontent.com/u/145979?s=50" width="50" height="50">](https://github.com/dstufft)<a name="note_26"></a>|@dstufft set owner to @glyph|
|-|-|

Ok coming back to this.

I think it looks good, my only suggestion is to add something like:

"Although ``inlineCallbacks`` looks similar to a synchronously written function, unlikely a synchronous function wherever a ``yield`` is used execution of the function will be paused while waiting for the result of the right hand side and other callbacks will be given a chance to execute in the mean time."

That will let people know without going into a ton of detail that their function execution will get paused and will hopefully be enough to give them some ideas if the notice issues from it about what to look for.

Other than that I think it looks great.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph set status to `closed`|
|-|-|

(In [43016]) Merge inlineCallbacks-howto-1009-2: Document inlineCallbacks

Author: glyph

Reviewer: dstufft

Fixes: [#7343](https://github.com/twisted/twisted/issues/7343)

.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph commented|
|-|-|

Thanks for the review, dstufft!
